### ISSUE TITLE
Defining metrics within loss functions was possible in Keras!

### ISSUE THREAD

**Issue by mvoelk** [tensorflow/tensorflow#25516]
**System information**
- TensorFlow version: 1.12.0
- Are you willing to contribute it: (Yes/No)

In Keras, it was possible to define metrics inside loss functions using closures. This means that something like the following 

```
class Loss():
    def __init__(self):
        self.metrics = []
    
    def loss(self, y_true, y_pred):
        
        # the following can be some complicated intermediate stuff that
        # would otherwise require redundant code
        t1 = tf.reduce_mean(y_true-y_pred)
        m2 = tf.reduce_max(y_true-y_pred)
        m3 = tf.reduce_min(y_true-y_pred)
        
        def m1(y_true, y_pred):
            return t1
        self.metrics.append(m1)
        
        def make_fcn(tensor, name):
            f = lambda y_true, y_pred: tensor
            f.__name__ = name
            return f
        for m in ['m2', 'm3']:
            self.metrics.append(make_fcn(eval(m),m))
        
        return tf.reduce_sum((y_true-y_pred)**2)

output2_loss = Loss()

model.compile(
    optimizer=keras.optimizers.RMSprop(lr=0.001, decay=1e-6), 
    loss={'pred': 'categorical_crossentropy', 'output2': output2_loss.loss},
    metrics={'pred': ['accuracy'], 'output2': output2_loss.metrics}
)
```

works perfectly fine in normal Keras but does not work in TensorFlow Keras.

I assume that there is no clear specification when the loss functions are evaluated and it is also clear to me that the approach may not work with eager execution.

Is there any elegant mechanism or workaround to achieve a similar behaviour?

**Comment by mvoelk** [comment]
The following works, but it is still more a hack than a solution.
```
class Loss():
    def __init__(self):
        self.metric_names = ['m1', 'm2', 'm3']
        
        def make_fcn(name):
            f = lambda y_true, y_pred: self._metric_functions[name](y_true, y_pred)
            f.__name__ = name
            return f
        self.metrics = [make_fcn(n) for n in self.metric_names]
        self._metric_functions = {}
    
    def loss(self, y_true, y_pred):
        
        # the following can be some complicated intermediate stuff that otherwise requires redundant code
        m1 = tf.reduce_mean(y_true-y_pred)
        m2 = tf.reduce_max(y_true-y_pred)
        m3 = tf.reduce_min(y_true-y_pred)
        
        for n in self.metric_names:
            t = eval(n)
            self._metric_functions[n] = lambda y_true, y_pred: t
        
        return tf.reduce_sum((y_true-y_pred)**2)

output2_loss = Loss()

model.compile(
    optimizer=keras.optimizers.RMSprop(lr=0.001, decay=1e-6), 
    loss={'pred': 'categorical_crossentropy', 'output2': output2_loss.loss},
    metrics={'pred': ['accuracy'], 'output2': output2_loss.metrics}
)
```

**Comment by pavithrasv** [comment]
Hello @mvoelk, definition of metric functions should not be dependent of when a loss function is called. This happens to work in keras-team/keras but it is definitely not an input that is supported.

Loss or metric inputs can be 
1. strings which are names of inbuilt functions
2. inbuilt functions
3. instances of inbuilt loss or metric classes
4. custom functions with standard (y_true, y_pred, ..) signature
5. instances of custom classes where the __call__ function has the standard signature.
